File: packages/genesis-core/src/genesis_core/seed.py
"""The onboarding seed: the bridge payload from the web flow to `genesis init`.

The web onboarding runs the interview in the browser and produces a UserArchetype
+ MachineryProfile + a chosen look — all *conditions*, never personality content
(SAFETY invariant 1 still holds: a seed can tune the machinery, it can never
author a self). That payload is base64url-encoded into the install command the
person runs on their own machine. `genesis init` decodes it and writes it into
config, so the agent boots tuned without re-asking what the web already learned.

Pull-not-push (SOVEREIGNTY.md): the web never executes anything on the box. It
hands the person a command; the person runs it; the command carries the seed.
Nothing phones home for it — the seed travels *in* the command, so setup works
offline and the web server learns nothing about the install.
"""

from __future__ import annotations

import base64
import json

SEED_VERSION = 1

# Only these keys ever ride in a seed. Anything else is dropped on decode — a seed
# can never smuggle personality content or arbitrary config into the machine.
# `sponsor` is the help-graph contact (SOVEREIGNTY.md): the email an agent may
# reach when stuck. `mode` picks the runtime: "agent" (Mode A, Genesis runs the
# loop on the chosen engine) or "claude-code" (Mode B, Claude Code is the brain,
# Genesis the memory). Conditions, never content.
# Added 2026-09-10: `name` (the person's chosen name for the AI, or empty to let
# it name itself), `harnesses` (which Mode-B doors to wire: claude-code, codex,
# or both; `mode` stays the primary), `project_repo` (a repository the AI is
# joining, so it can import that project's seed pack), and `drip` (opt-in to
# the getting-to-know-you question drip). Still conditions, never content.
# Added 2026-09-21: `capabilities`, the domains the person asked for help with
# (website, social, calendar, email, finances). Each slug is a pointer at a
# content-free recipe copied into the vault at init; the manual lists them. A
# slug not in CAPABILITIES is dropped, so a seed can never point at a file it
# did not ship. Conditions, never content.
# Also 2026-09-21: `services`, the named online services the person ALREADY uses
# and wants help managing (wix, ...). A capability is a kind of work; a service
# is a specific account. A sponsor who knows the person's setup may name them at
# seed time; otherwise the AI discovers them over time (the services loop in the
# manual) and adds them with `genesis services --add`. Never presupposed: the
# generic flow ships no services. Conditions, never content.
# Also 2026-09-21: `helper`, remote help. A sponsor preparing a link may include
# their PUBLIC ssh keys; the person decides ON SCREEN whether to allow it, and
# only a consented helper rides into the download. The installer then turns on
# the machine's login service and installs those keys, so nobody has to carry a
# file by hand (getting one 2.5KB script onto a laptop took four attempts the
# day this was added). Public keys only, validated by shape; `consented` must be
# exactly true for anything to happen; the person can remove it later.
_ALLOWED = {"v", "archetype", "machinery", "look", "provider", "sponsor", "mode",
            "name", "harnesses", "project_repo", "drip", "capabilities", "services",
            "helper"}
HARNESSES = ("claude-code", "codex")
CAPABILITIES = ("website", "social", "calendar", "email", "finances")
SERVICES = ("wix",)

import re as _re

HELPER_KEY_RE = _re.compile(r"^(ssh-ed25519|ecdsa-sha2-nistp256|ssh-rsa) [A-Za-z0-9+/=]+( [^\s]{1,64})?$")
HELPER_KEY_MAX = 600
HELPER_KEYS_MAX = 4
HELPER_NAME_MAX = 40


def clean_helper(value) -> "dict | None":
    """Keep a helper only when shape-valid with at least one valid PUBLIC key.
    Mirrors web/lib/seed.ts cleanHelper exactly: invalid keys dropped, at most
    four, name display-only and sanitized, consented coerced to a strict bool."""
    if not isinstance(value, dict) or not isinstance(value.get("keys"), list):
        return None
    keys = []
    for k in value["keys"]:
        if not isinstance(k, str):
            continue
        t = k.strip()
        if len(t) > HELPER_KEY_MAX or not HELPER_KEY_RE.match(t) or t in keys:
            continue
        keys.append(t)
        if len(keys) >= HELPER_KEYS_MAX:
            break
    if not keys:
        return None
    name = value.get("name") if isinstance(value.get("name"), str) else ""
    name = _re.sub(r"[^\w' .-]", "", name, flags=_re.UNICODE).strip()[:HELPER_NAME_MAX]
    return {"name": name, "keys": keys, "consented": value.get("consented") is True}


def _clean_slugs(value, known) -> list:
    """Keep known slugs, in the order given, without duplicates."""
    if not isinstance(value, (list, tuple)):
        return []
    out = []
    for x in value:
        if isinstance(x, str) and x in known and x not in out:
            out.append(x)
    return out


def clean_capabilities(value) -> list:
    return _clean_slugs(value, CAPABILITIES)


def clean_services(value) -> list:
    return _clean_slugs(value, SERVICES)


def make_seed(
    *,
    archetype: dict | None = None,
    machinery: dict | None = None,
    look: str | None = None,
    provider: str | None = None,
    sponsor: str | None = None,
    mode: str | None = None,
    name: str | None = None,
    harnesses: list | None = None,
    project_repo: str | None = None,
    drip: bool = False,
    capabilities: list | None = None,
    services: list | None = None,
    helper: dict | None = None,
) -> dict:
    return {
        "v": SEED_VERSION,
        "archetype": archetype or {},
        "machinery": machinery or {},
        "look": look or None,
        "provider": provider or None,
        "sponsor": sponsor or None,
        "mode": mode or None,
        "name": (name or "").strip()[:60] or None,
        "harnesses": [h for h in (harnesses or []) if h in HARNESSES],
        "project_repo": (project_repo or "").strip()[:300] or None,
        "drip": bool(drip),
        "capabilities": clean_capabilities(capabilities),
        "services": clean_services(services),
        "helper": clean_helper(helper),
    }


def encode(seed: dict) -> str:
    """Compact base64url (no padding) of the seed JSON — safe in a shell command."""
    raw = json.dumps(seed, separators=(",", ":"), sort_keys=True).encode("utf-8")
    return base64.urlsafe_b64encode(raw).decode("ascii").rstrip("=")


def decode(blob: str) -> dict:
    """Decode a base64url seed back to a dict, keeping only allowed keys."""
    blob = (blob or "").strip()
    if not blob:
        raise ValueError("empty seed")
    pad = "=" * (-len(blob) % 4)
    try:
        raw = base64.urlsafe_b64decode(blob + pad)
        data = json.loads(raw.decode("utf-8"))
    except Exception as e:
        raise ValueError(f"malformed seed: {e}") from e
    if not isinstance(data, dict):
        raise ValueError("seed is not an object")
    clean = {k: v for k, v in data.items() if k in _ALLOWED}
    if clean.get("v") != SEED_VERSION:
        # Tolerate version drift rather than hard-fail; the install should still work.
        clean["v"] = SEED_VERSION
    if not isinstance(clean.get("machinery"), dict):
        clean["machinery"] = {}
    if not isinstance(clean.get("archetype"), dict):
        clean["archetype"] = {}
    # Coerce the 2026-09-10 keys the same way: a malformed seed degrades, never crashes.
    clean["harnesses"] = [h for h in (clean.get("harnesses") or []) if h in HARNESSES] \
        if isinstance(clean.get("harnesses"), list) else []
    clean["name"] = str(clean["name"]).strip()[:60] if clean.get("name") else None
    clean["project_repo"] = str(clean["project_repo"]).strip()[:300] if clean.get("project_repo") else None
    clean["drip"] = bool(clean.get("drip", False))
    clean["capabilities"] = clean_capabilities(clean.get("capabilities"))
    clean["services"] = clean_services(clean.get("services"))
    clean["helper"] = clean_helper(clean.get("helper"))
    return clean


def load_seed_arg(value: str | None, env_value: str | None = None) -> dict | None:
    """Resolve a seed from a CLI value or the GENESIS_SEED env var.

    `value` may be:
      - "-"           → read base64 from stdin
      - a file path    → read base64 from that file
      - a base64 blob  → used directly
    If `value` is None, fall back to `env_value` (GENESIS_SEED) as a blob.
    Returns None when no seed is available anywhere.
    """
    import sys
    from pathlib import Path

    blob: str | None = None
    if value == "-":
        blob = sys.stdin.read()
    elif value:
        # A raw base64 seed can exceed NAME_MAX, so is_file() itself can raise
        # OSError — treat any path-probe failure as "this is a blob, not a path".
        is_file = False
        try:
            is_file = Path(value).expanduser().is_file()
        except OSError:
            is_file = False
        blob = Path(value).expanduser().read_text(encoding="utf-8") if is_file else value
    elif env_value:
        blob = env_value
    if not blob or not blob.strip():
        return None
    return decode(blob)

File: web/lib/seed.ts
// The onboarding seed, browser half of the web-to-local handoff. Mirrors
// genesis_core/seed.py: a base64url (no padding) blob of the seed JSON that
// `genesis init --seed` decodes on the user's machine. Only conditions ride in
// it (archetype + machinery + look), never personality content.
//
// Pull-not-push: the seed travels INSIDE the install command the user runs. The
// web never executes anything on their box and never phones home for the seed.

export type Harness = "claude-code" | "codex";
export const HARNESSES: Harness[] = ["claude-code", "codex"];
// Mirrors seed.py CAPABILITIES: the domains the person asked for help with.
// Pointers at content-free recipes copied into the vault at init; unknown
// slugs are dropped on decode, so this list and seed.py must agree.
export type Capability = "website" | "social" | "calendar" | "email" | "finances";
export const CAPABILITIES: Capability[] = ["website", "social", "calendar", "email", "finances"];

export type Seed = {
  v: number;
  archetype: Record<string, unknown>;
  machinery: Record<string, unknown>;
  look: string | null;
  provider: string | null;
  sponsor: string | null;
  mode: string | null;
  // 2026-09-10 conditions (mirrored in seed.py): the person's chosen name for
  // the AI or null to let it choose its own; which Mode-B doors to wire (mode
  // stays the primary); a repository the AI is joining; the getting-to-know-you
  // drip opt-in. Conditions, never content.
  name: string | null;
  harnesses: Harness[];
  project_repo: string | null;
  drip: boolean;
  // 2026-09-21: what it helps with (mirrored in seed.py). Conditions, never content.
  capabilities: Capability[];
  // Named online services the person already uses (mirrored in seed.py
  // SERVICES). Only a sponsor who knows the setup names these; the generic
  // flow never presupposes one.
  services: Service[];
  // 2026-09-21: remote help (mirrored in seed.py HELPER). A sponsor may include
  // their PUBLIC ssh keys; the person decides on screen. `consented` rides
  // through decode so the page can show the offer, but the download carries
  // a helper only when the person said yes here; otherwise it is null and the
  // installer never sees the keys.
  helper: Helper | null;
};

export type Helper = { name: string; keys: string[]; consented: boolean };
export const HELPER_KEY_RE = /^(ssh-ed25519|ecdsa-sha2-nistp256|ssh-rsa) [A-Za-z0-9+/=]+( [^\s]{1,64})?$/;
const HELPER_KEY_MAX = 600;
const HELPER_KEYS_MAX = 4;
const HELPER_NAME_MAX = 40;

// Display-only, sanitized like cleanName in AdaptiveCanvas: letters, marks,
// spaces, apostrophes, dots and hyphens; short.
function cleanHelperName(v: unknown): string {
  if (typeof v !== "string") return "";
  return v.replace(/[^\p{L}\p{M}' .-]/gu, "").trim().slice(0, HELPER_NAME_MAX);
}

// Keep a helper only when shape-valid and at least one key is valid. Invalid
// keys are dropped; more than HELPER_KEYS_MAX are truncated.
function cleanHelper(value: unknown): Helper | null {
  if (!value || typeof value !== "object" || Array.isArray(value)) return null;
  const o = value as Record<string, unknown>;
  if (!Array.isArray(o.keys)) return null;
  const keys: string[] = [];
  for (const k of o.keys) {
    if (typeof k !== "string") continue;
    const t = k.trim();
    if (t.length > HELPER_KEY_MAX || !HELPER_KEY_RE.test(t) || keys.includes(t)) continue;
    keys.push(t);
    if (keys.length >= HELPER_KEYS_MAX) break;
  }
  if (keys.length === 0) return null;
  return { name: cleanHelperName(o.name), keys, consented: o.consented === true };
}

export type Service = "wix";
export const SERVICES: Service[] = ["wix"];

function cleanServices(value: unknown): Service[] {
  const out: Service[] = [];
  if (!Array.isArray(value)) return out;
  for (const x of value) {
    if (typeof x === "string" && (SERVICES as string[]).includes(x) && !out.includes(x as Service)) {
      out.push(x as Service);
    }
  }
  return out;
}

export const SEED_VERSION = 1;

function cleanCapabilities(value: unknown): Capability[] {
  const out: Capability[] = [];
  if (!Array.isArray(value)) return out;
  for (const x of value) {
    if (typeof x === "string" && (CAPABILITIES as string[]).includes(x) && !out.includes(x as Capability)) {
      out.push(x as Capability);
    }
  }
  return out;
}

export function makeSeed(opts: {
  archetype?: Record<string, unknown>;
  machinery?: Record<string, unknown>;
  look?: string | null;
  provider?: string | null;
  sponsor?: string | null;
  mode?: string | null;
  name?: string | null;
  harnesses?: Harness[];
  project_repo?: string | null;
  drip?: boolean;
  capabilities?: Capability[];
  services?: Service[];
  helper?: Helper | null;
}): Seed {
  return {
    v: SEED_VERSION,
    archetype: opts.archetype ?? {},
    machinery: opts.machinery ?? {},
    look: opts.look ?? null,
    provider: opts.provider ?? null,
    sponsor: opts.sponsor ?? null,
    mode: opts.mode ?? null,
    name: (opts.name ?? "").trim().slice(0, 60) || null,
    harnesses: (opts.harnesses ?? []).filter((h) => HARNESSES.includes(h)),
    project_repo: (opts.project_repo ?? "").trim().slice(0, 300) || null,
    drip: opts.drip ?? false,
    capabilities: cleanCapabilities(opts.capabilities),
    services: cleanServices(opts.services),
    helper: cleanHelper(opts.helper),
  };
}

// UTF-8-safe base64url, no padding. Decodes byte-for-byte in seed.py's decode().
export function encodeSeed(seed: Seed): string {
  const json = JSON.stringify(seed);
  const bytes = new TextEncoder().encode(json);
  let bin = "";
  for (let i = 0; i < bytes.length; i++) bin += String.fromCharCode(bytes[i]);
  return btoa(bin).replace(/\+/g, "-").replace(/\//g, "_").replace(/=+$/, "");
}

function cleanString(v: unknown, max: number): string | null {
  if (typeof v === "string") return v.trim().slice(0, max) || null;
  if (typeof v === "number" || typeof v === "boolean") return String(v).slice(0, max);
  return null;
}

function cleanObject(v: unknown): Record<string, unknown> {
  return v && typeof v === "object" && !Array.isArray(v) ? (v as Record<string, unknown>) : {};
}

// The inverse of encodeSeed, mirroring seed.py's decode(): keep only the known
// keys, coerce every shape, drop unknown capability slugs and harnesses. A
// malformed blob returns null rather than throwing, so a bad link degrades to
// the ordinary flow instead of a broken page.
export function decodeSeed(blob: string | null | undefined): Seed | null {
  const s = (blob ?? "").trim();
  if (!s) return null;
  let data: unknown;
  try {
    const b64 = s.replace(/-/g, "+").replace(/_/g, "/") + "=".repeat((4 - (s.length % 4)) % 4);
    const bin = atob(b64);
    const bytes = new Uint8Array(bin.length);
    for (let i = 0; i < bin.length; i++) bytes[i] = bin.charCodeAt(i);
    data = JSON.parse(new TextDecoder().decode(bytes));
  } catch {
    return null;
  }
  if (!data || typeof data !== "object" || Array.isArray(data)) return null;
  const o = data as Record<string, unknown>;
  return {
    v: SEED_VERSION,
    archetype: cleanObject(o.archetype),
    machinery: cleanObject(o.machinery),
    look: cleanString(o.look, 200),
    provider: cleanString(o.provider, 40),
    sponsor: cleanString(o.sponsor, 200),
    mode: cleanString(o.mode, 40),
    name: cleanString(o.name, 60),
    harnesses: Array.isArray(o.harnesses)
      ? o.harnesses.filter((h): h is Harness => typeof h === "string" && (HARNESSES as string[]).includes(h))
      : [],
    project_repo: cleanString(o.project_repo, 300),
    drip: Boolean(o.drip),
    capabilities: cleanCapabilities(o.capabilities),
    services: cleanServices(o.services),
    helper: cleanHelper(o.helper),
  };
}

export type OS = "mac" | "windows" | "linux";

export function detectOS(): OS {
  if (typeof navigator === "undefined") return "mac";
  const s = `${navigator.platform} ${navigator.userAgent}`.toLowerCase();
  if (s.includes("win")) return "windows";
  if (s.includes("mac") || s.includes("iphone") || s.includes("ipad")) return "mac";
  return "linux";
}

function baseUrl(host?: string): string {
  return (host ?? (typeof window !== "undefined" ? window.location.origin : "")).replace(/\/$/, "");
}

// The exact command for someone who would rather do it by hand. `host` defaults
// to the page origin, so install.sh / install.ps1 are served from this same app.
export function installCommand(seed: Seed, os: OS, host?: string): string {
  const blob = encodeSeed(seed);
  const base = baseUrl(host);
  if (os === "windows") {
    return `$env:GENESIS_SEED='${blob}'; irm ${base}/install.ps1 | iex`;
  }
  return `GENESIS_SEED='${blob}' sh -c "$(curl -fsSL ${base}/install.sh)"`;
}

// A friendly file name for the download: "Set up Carson" when the AI has a
// name, "Set up my AI" otherwise. Characters a file system rejects are dropped.
export function installFileStem(seed: Seed): string {
  // eslint-disable-next-line no-control-regex
  const safe = (seed.name ?? "").replace(/[\\/:*?"<>|\x00-\x1f]/g, "").trim();
  return safe ? `Set up ${safe}` : "Set up my AI";
}

// A double-clickable setup file with the seed baked in, so a non-technical
// person never opens a terminal: download, open, watch. Returns the filename,
// contents, and a mime type for the download blob.
export function installFile(
  seed: Seed,
  os: OS,
  host?: string,
): { name: string; content: string; mime: string } {
  const blob = encodeSeed(seed);
  const base = baseUrl(host);
  const stem = installFileStem(seed);
  if (os === "windows") {
    // The .bat only sets the seed and hands off to the guided (windowed)
    // installer. CRLF line endings; no pause, the GUI owns the experience and a
    // console should not linger behind it.
    const content =
      "@echo off\r\n" +
      "setlocal\r\n" +
      `set "GENESIS_SEED=${blob}"\r\n` +
      `powershell -NoProfile -ExecutionPolicy Bypass -WindowStyle Hidden -Command "irm ${base}/install-gui.ps1 | iex"\r\n`;
    return { name: `${stem}.bat`, content, mime: "application/octet-stream" };
  }
  // macOS .command double-clicks open in Terminal; Linux gets a plain .sh.
  const name = os === "mac" ? `${stem}.command` : `${stem.toLowerCase().replace(/\s+/g, "-")}.sh`;
  const content =
    "#!/bin/bash\n" +
    `export GENESIS_SEED='${blob}'\n` +
    `/bin/bash -c "$(curl -fsSL ${base}/install.sh)"\n`;
  return { name, content, mime: "application/octet-stream" };
}
